lib/rules/valid-typeof.js

```diff
@@ -72,6 +72,8 @@ module.exports = {
                             if (VALID_TYPES.indexOf(value) === -1) {
                                 context.report({ node: sibling, messageId: "invalidValue" });
                             }
+                        } else if (sibling.type === "Identifier" && sibling.name === "undefined") {
+                            context.report({ node: sibling, messageId: "notString" });
```

Comment:
It seems wrong to report `Typeof comparisons should be to string literals.` unless `{ "requireStringLiterals": true }` is set. I think the message for this should be `"invalidValue"`, because the value is the cause of the problem here, and that this check should be after `if (requireStringLiterals && !isTypeofExpression(sibling)`, so that it works like this:

```js
/* eslint valid-typeof: ["error", { "requireStringLiterals": true }] */

typeof x === undefined; //  error  Typeof comparisons should be to string literals  valid-typeof
```

```js
/* eslint valid-typeof: ["error"] */

typeof x === undefined; //  error  Invalid typeof comparison value  valid-typeof

```

Reply:
We should also check if the identifier is a reference to the global `undefined` variable.

```js
/* eslint valid-typeof: ["error"] */

(undefined) => {
    typeof x === undefined; // false positive
}
```

An example of how to do that is [`function isReferenceToGlobalVariable`](https://github.com/eslint/eslint/blob/5d65c3bc1e514ed07406c502437a1642913b27ed/lib/rules/id-denylist.js#L143) in `id-denylist` rule.